worldedit-core/src/main/java/com/fastasyncworldedit/core/util/task/KeyQueuedExecutorService.java

```diff
@@ -0,0 +1,164 @@
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Executor service that queues tasks based on keys, executing tasks on a configurable {@link ThreadPoolExecutor}
+ *
+ * @param <K> Key type
+ */
+public class KeyQueuedExecutorService<K> {
+
+    private final ExecutorService parent;
+    private final Map<K, KeyRunner> keyQueue = new HashMap<>();
+
+    /**
+     * Create a new {@link KeyQueuedExecutorService} instance
+     *
+     * @param parent Parent {@link ExecutorService} to use for actual task completion
+     */
+    public KeyQueuedExecutorService(ExecutorService parent) {
+        this.parent = parent;
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#shutdown()}
+     */
+    public void shutdown() {
+        parent.shutdown();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#shutdownNow()}
+     */
+    @Nonnull
+    public List<Runnable> shutdownNow() {
+        return parent.shutdownNow();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#isShutdown()}
+     */
+    public boolean isShutdown() {
+        return parent.isShutdown();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#isTerminated()}
+     */
+    public boolean isTerminated() {
+        return parent.isTerminated();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#awaitTermination(long, TimeUnit)}
+     */
+    public boolean awaitTermination(long timeout, @Nonnull TimeUnit unit) throws InterruptedException {
+        return parent.awaitTermination(timeout, unit);
+    }
+
+    protected <T> FutureTask<T> newTaskFor(Runnable runnable, T value) {
+        return new FutureTask<>(runnable, value);
+    }
+
+    protected <T> FutureTask<T> newTaskFor(Callable<T> callable) {
+        return new FutureTask<>(callable);
+    }
+
+    @Nonnull
+    public <T> Future<T> submit(@Nonnull K key, @Nonnull Callable<T> task) {
+        FutureTask<T> ftask = newTaskFor(task);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    @Nonnull
+    public <T> Future<T> submit(@Nonnull K key, @Nonnull Runnable task, T result) {
+        FutureTask<T> ftask = newTaskFor(task, result);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    @Nonnull
+    public Future<?> submit(@Nonnull K key, @Nonnull Runnable task) {
+        FutureTask<Void> ftask = newTaskFor(task, null);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    public void execute(@Nonnull K key, @Nonnull FutureTask<?> command) {
+        KeyRunner runner;
+        synchronized (keyQueue) {
+            runner = keyQueue.computeIfAbsent(key, KeyRunner::new);
+            runner.add(command);
+            runner.triggerRun();
```

Comment:
If I see this right, this might be called while the `command` task was already processed by a previous call to `triggerRun`, causing an exception. (and alternatively, run multiple commands for one player concurrently)

I think we need some logic to ensure only one worker is draining the queue here